File: server/datasets/python_source-code/Multiple-E1/organic_set_1_grid_dfs.py
def numIslands(grid):
    if not grid: return 0
    visited = set()
    count = 0

    def traverse(r, c):
        if r not in range(len(grid)) or c not in range(len(grid[0])) or grid[r][c] == "0" or (r, c) in visited:
            return
        visited.add((r, c))
        traverse(r + 1, c)
        traverse(r - 1, c)
        traverse(r, c + 1)
        traverse(r, c - 1)

    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == "1" and (r, c) not in visited:
                count += 1
                traverse(r, c)
    return count

File: server/datasets/java_source-code/Multiple-E1/organic_set_1_grid_dfs.java
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        Set<String> visited = new HashSet<>();
        int count = 0;

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (grid[r][c] == '1' && !visited.contains(r + "," + c)) {
                    count++;
                    traverse(grid, r, c, visited);
                }
            }
        }
        return count;
    }

    private void traverse(char[][] grid, int r, int c, Set<String> visited) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length || grid[r][c] == '0' || visited.contains(r + "," + c)) {
            return;
        }
        visited.add(r + "," + c);
        traverse(grid, r + 1, c, visited);
        traverse(grid, r - 1, c, visited);
        traverse(grid, r, c + 1, visited);
        traverse(grid, r, c - 1, visited);
    }
}
